rasterio/rio/shapes.py

```diff
@@ -110,134 +106,147 @@ def shapes(
-                    img = msk
-
-                # Read the band data unless the --mask option is given.
-                if band:
-                    if sampling == 1:
-                        img = src.read(bidx, masked=False)
-                    else:
-                        img = np.zeros(
-                            shape,
-                            dtype=src.dtypes[src.indexes.index(bidx)])
-                        img = src.read(bidx, img, masked=False)
-
-                # If --as-mask option was given, convert the image
-                # to a binary image. This reduces the number of shape
-                # categories to 2 and likely reduces the number of
-                # shapes.
-                if as_mask:
-                    tmp = np.ones_like(img, 'uint8') * 255
-                    tmp[img == 0] = 0
-                    img = tmp
-                    if not with_nodata:
-                        msk = tmp
-
-                # Transform the raster bounds.
-                bounds = src.bounds
-                xs = [bounds[0], bounds[2]]
-                ys = [bounds[1], bounds[3]]
-                if projection == 'geographic':
-                    xs, ys = rasterio.warp.transform(
-                        src.crs, CRS({'init': 'epsg:4326'}), xs, ys)
-                if precision >= 0:
-                    xs = [round(v, precision) for v in xs]
-                    ys = [round(v, precision) for v in ys]
-                self._xs = xs
-                self._ys = ys
-
-                # Prepare keyword arguments for shapes().
-                kwargs = {'transform': transform}
-                if not with_nodata:
-                    kwargs['mask'] = msk
-
-                src_basename = os.path.basename(src.name)
-
-                # Yield GeoJSON features.
-                for i, (g, val) in enumerate(
-                        rasterio.features.shapes(img, **kwargs)):
-                    if projection == 'geographic':
-                        g = rasterio.warp.transform_geom(
-                            src.crs, 'EPSG:4326', g,
-                            antimeridian_cutting=True, precision=precision)
-                    xs, ys = zip(*coords(g))
-                    yield {
-                        'type': 'Feature',
-                        'id': "{0}:{1}".format(src_basename, i),
-                        'properties': {
-                            'val': val, 'filename': src_basename
-                        },
-                        'bbox': [min(xs), min(ys), max(xs), max(ys)],
-                        'geometry': g
-                    }
-
     if not sequence:
         geojson_type = 'collection'
 
     try:
         with ctx.obj['env'] as env:
             write_features(
-                stdout, Collection(env), sequence=sequence,
+                stdout,
+                feature_gen(
+                    input, env, bidx, sampling,  band, as_mask, with_nodata, projection, precision),
```

Comment:
The last four arguments to `feature_gen` feel like keyword args to me instead of positional args.